{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "Missing parentheses in call to 'print'. Did you mean print(\"\"\"\n{usage}\nParameters\n----------\nThis script will generate {args.nfeatures} random features, each about {size}\nbp, from genomic coordinates {args.chrom}:{args.start}-{args.stop}.\n\nGenomic signal will be extracted from these files:\n  {files}\n\nFor each file, the signal for the {args.nfeatures} features will be extracted\nin parallel and binned into {args.bins} bins, using from 1 to {max_proc} CPUs.\nEach CPU will get as many as {args.chunksize} features at a time.\n\nPlot will be saved as {plot_filename}.\n\"\"\")? (<ipython-input-4-c235221e3a78>, line 110)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-4-c235221e3a78>\"\u001b[0;36m, line \u001b[0;32m110\u001b[0m\n\u001b[0;31m    \"\"\"\u001b[0m\n\u001b[0m       \n^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m Missing parentheses in call to 'print'. Did you mean print(\"\"\"\n{usage}\nParameters\n----------\nThis script will generate {args.nfeatures} random features, each about {size}\nbp, from genomic coordinates {args.chrom}:{args.start}-{args.stop}.\n\nGenomic signal will be extracted from these files:\n  {files}\n\nFor each file, the signal for the {args.nfeatures} features will be extracted\nin parallel and binned into {args.bins} bins, using from 1 to {max_proc} CPUs.\nEach CPU will get as many as {args.chunksize} features at a time.\n\nPlot will be saved as {plot_filename}.\n\"\"\")?\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "usage=\"\"\"\n",
    "This is a benchmarking script for the metaseq Python package\n",
    "(https://github.com/daler/metaseq).  It can help decide how many processors are\n",
    "optimal to use for your particular hardware configuration.\n",
    "\n",
    "The default arguments work for data shipped along with metaseq.  The default\n",
    "arguments generate 1000 random features of about 5kb in size and then extract\n",
    "signal data for BED, BAM, bigWig, and bigBed formats. This is done using from\n",
    "1 to however many CPUs are availble.\n",
    "\n",
    "The result is a plot showing the actual time in seconds per 1000 features, over\n",
    "a range of processors.  A dotted line for the optimal 1/x speedup.  The plot is\n",
    "also saved as a PDF.\n",
    "\n",
    "Note that results are a function of CPU speed, disk I/O throughput, and nature\n",
    "of the data (for BAM, BED, and bigBed formats), so your mileage may vary.  For\n",
    "example, multiple CPUs will not help if there's only a small number of\n",
    "features.\n",
    "\"\"\"\n",
    "import sys\n",
    "import numpy as np\n",
    "import time\n",
    "from matplotlib import pyplot as plt\n",
    "import pybedtools\n",
    "import metaseq\n",
    "import multiprocessing\n",
    "import argparse\n",
    "import os\n",
    "\n",
    "ap = argparse.ArgumentParser(usage=usage)\n",
    "ap.add_argument(\n",
    "    '-n', '--nfeatures', type=int, default=1000,\n",
    "    help='Number of random windows to make.  The size of each feature is '\n",
    "    '(start - stop) / nfeatures, so changing this parameter will affect '\n",
    "    'the size of each feature.')\n",
    "ap.add_argument(\n",
    "    '-c', '--chunksize', type=int, default=100,\n",
    "    help='Number of windows to send to each process')\n",
    "ap.add_argument(\n",
    "    '--start', type=int, default=10000,\n",
    "    help='start coord of possible windows. See --nfeatures help regarding size.')\n",
    "ap.add_argument(\n",
    "    '--stop', type=int, default=5000000,\n",
    "    help='stop coord of possible windows. See --nfeatures help regarding size.')\n",
    "ap.add_argument(\n",
    "    '--chrom', default='chr2L',\n",
    "    help='chromsome to make windows on')\n",
    "ap.add_argument(\n",
    "    '--type', default='all',\n",
    "    help='Only use the specified file types.  Either \"all\" '\n",
    "    '(default) or a comma-separated list of [bam, bigwig, bed, '\n",
    "    'bigbed].')\n",
    "ap.add_argument(\n",
    "    '--prefix', default=os.path.join(metaseq.data_dir(), 'x'),\n",
    "    help='Prefix of filenames to use.  Expects files '\n",
    "    'with this prefix, and the following suffixes: .bam, .bigwig, '\n",
    "    '.bed.gz (should already be tabixed), .bigbed.  Default: %(default)s')\n",
    "ap.add_argument(\n",
    "    '--plot-prefix', default='./speedtest',\n",
    "    help='Filename used to save the resulting plot. Default is %(default)s')\n",
    "ap.add_argument(\n",
    "    '--bins', default=100,\n",
    "    help='Number of bins for each feature')\n",
    "args = ap.parse_args()\n",
    "\n",
    "\n",
    "requested = args.type.split(',')\n",
    "allowed = ['bam', 'bed', 'bigwig', 'bigbed', 'all']\n",
    "for req in requested:\n",
    "    if req not in allowed:\n",
    "        raise ValueError(\"%s not in %s\" % (req, allowed))\n",
    "\n",
    "plt.rcParams['font.size'] = 10\n",
    "\n",
    "intervals = pybedtools.BedTool().window_maker(\n",
    "    genome={args.chrom: (args.start, args.stop)}, n=args.nfeatures)\\\n",
    "    .shuffle(genome={args.chrom: (args.start, args.stop)}, seed=1)\n",
    "\n",
    "size = (args.stop - args.start) / args.nfeatures\n",
    "if args.type == 'all':\n",
    "    requested = ['bigwig', 'bam', 'bigbed', 'bed']\n",
    "signals = []\n",
    "for req in requested:\n",
    "    if req != 'bed':\n",
    "        signals.append(metaseq.genomic_signal(args.prefix + '.' + req, req))\n",
    "    else:\n",
    "        signals.append(metaseq.genomic_signal(args.prefix + '.bed.gz', 'bed'))\n",
    "\n",
    "files = '* ' + '\\n  * '.join([i.fn for i in signals])\n",
    "plot_filename = (args.plot_prefix + '-%s_features-%s_bp_chunksize=%s.pdf'\n",
    "                 % (args.nfeatures, size, args.chunksize))\n",
    "processes = range(1, multiprocessing.cpu_count() + 1)\n",
    "max_proc = processes[-1]\n",
    "print \"\"\"\n",
    "{usage}\n",
    "Parameters\n",
    "----------\n",
    "This script will generate {args.nfeatures} random features, each about {size}\n",
    "bp, from genomic coordinates {args.chrom}:{args.start}-{args.stop}.\n",
    "\n",
    "Genomic signal will be extracted from these files:\n",
    "  {files}\n",
    "\n",
    "For each file, the signal for the {args.nfeatures} features will be extracted\n",
    "in parallel and binned into {args.bins} bins, using from 1 to {max_proc} CPUs.\n",
    "Each CPU will get as many as {args.chunksize} features at a time.\n",
    "\n",
    "Plot will be saved as {plot_filename}.format(**locals())\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "d = {}\n",
    "results = {}\n",
    "for x in signals:\n",
    "    print '\\n' + x.kind, '[%s]' % x.fn\n",
    "    sys.stdout.flush()\n",
    "    times = []\n",
    "    print 'CPUs:',\n",
    "    for p in processes:\n",
    "        print p,\n",
    "        sys.stdout.flush()\n",
    "        if p == 1:\n",
    "            p = None\n",
    "        t = []\n",
    "        t0 = time.time()\n",
    "        a = x.array(intervals, bins=args.bins, processes=p,\n",
    "                    chunksize=args.chunksize)\n",
    "        elapsed = time.time() - t0\n",
    "        times.append(elapsed)\n",
    "    results[x.kind] = a.sum(axis=0)\n",
    "    d[x.kind] = times\n",
    "\n",
    "fig = plt.figure(figsize=(12, 6))\n",
    "i = 1\n",
    "for k, t in d.items():\n",
    "    ax = fig.add_subplot(2, len(signals), i)\n",
    "    t = np.array(t) / (len(intervals) / 1000.)\n",
    "    ax.plot(processes, t, 'k.-', label=k)\n",
    "    ax.plot(processes, 1 / np.array(processes, dtype=float) * t[0], 'k:')\n",
    "    ax.set_title(k)\n",
    "    plt.xlabel('# processes')\n",
    "    plt.ylabel('time per 1k windows (s)')\n",
    "\n",
    "    ax = fig.add_subplot(2, len(signals), i + len(signals))\n",
    "    ax.plot(results[k], 'k')\n",
    "    ax.set_ylabel('reads\\nper bin')\n",
    "    ax.set_xlabel('bins')\n",
    "    i += 1\n",
    "fig.subplots_adjust(left=0.1, right=0.95, wspace=0.5, hspace=0.5)\n",
    "fig.suptitle('%s features, %sbp each, chunksize=%s' % (args.nfeatures, size, args.chunksize))\n",
    "fig.savefig(plot_filename)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
