{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-2-7b6f831b3cef>, line 540)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-2-7b6f831b3cef>\"\u001b[0;36m, line \u001b[0;32m540\u001b[0m\n\u001b[0;31m    assert gs['bam'].genome() == {'chr2L': (0, 23011544L)}\u001b[0m\n\u001b[0m                                                       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Many of these tests use the minimal test/data/gdc.bed file which has just\n",
    "enough complexity to be useful in testing corner cases.  When reading through\n",
    "the tests, it's useful to have that file open to understand what's happening.\n",
    "\"\"\"\n",
    "import os\n",
    "import metaseq\n",
    "import multiprocessing\n",
    "from metaseq.array_helpers import ArgumentError\n",
    "import numpy as np\n",
    "from nose.tools import assert_raises\n",
    "from nose.plugins.skip import SkipTest\n",
    "\n",
    "nan = np.nan\n",
    "inf = np.inf\n",
    "\n",
    "gs = {}\n",
    "for kind in ['bed', 'bam', 'bigbed', 'bigwig']:\n",
    "    gs[kind] = metaseq.genomic_signal(metaseq.example_filename('gdc.%s' % kind), kind)\n",
    "\n",
    "PROCESSES = int(os.environ.get(\"METASEQ_PROCESSES\", multiprocessing.cpu_count()))\n",
    "\n",
    "def test_tointerval():\n",
    "    assert metaseq.helpers.tointerval(\"chr2L:1-10[-]\").strand == '-'\n",
    "    assert metaseq.helpers.tointerval(\"chr2L:1-10[+]\").strand == '+'\n",
    "    assert metaseq.helpers.tointerval(\"chr2L:1-10\").strand == '.'\n",
    "\n",
    "\n",
    "def test_local_count():\n",
    "\n",
    "    def check(kind, coord, expected, stranded):\n",
    "        try:\n",
    "            result = gs[kind].local_count(coord, stranded=stranded)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert result == expected, (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed']:\n",
    "        for coord, expected, stranded in (\n",
    "            ('chr2L:1-80', 3, False),       #  easy case\n",
    "            ('chr2L:1000-3000', 0, False),  #  above upper boundary\n",
    "            ('chr2L:1-9', 0, False),        #  below lower boundary\n",
    "            ('chr2L:71-73[-]', 2, False),   #  unstranded = 2\n",
    "            ('chr2L:71-73[-]', 1, True),    #  stranded = 1\n",
    "            ('chr2L:70-71', 2, False),      #  pathological corner case\n",
    "            # ('chr2L:75-76', 0, False),      #  pathological corner case\n",
    "        ):\n",
    "            yield check, kind, coord, expected, stranded\n",
    "\n",
    "\n",
    "def test_local_coverage_stranded():\n",
    "    def check(kind, coord, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            ('chr2L:1-20[-]',\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0.])[::-1],\n",
    "                 # note reverse------------------------------------------------------------------------^^^^^^\n",
    "             ),\n",
    "            ),\n",
    "            ('chr2L:68-76[-]',\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 0., 2., 2., 2., 2., 2., 0.])[::-1],\n",
    "                 # note reverse----------------------------^^^^^^\n",
    "\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            yield check, kind, coord, expected\n",
    "\n",
    "\n",
    "def test_local_coverage_shifted():\n",
    "    def check(kind, coord, shift_width, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord, shift_width=shift_width)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed']:\n",
    "        for coord, shift_width, expected in (\n",
    "            ('chr2L:1-20', -2,\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "            # this one is complex, because the minus-strand read shifts left,\n",
    "            # and the plus-strand shifts right.\n",
    "            ('chr2L:68-76', 1,\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 1., 1., 2., 2., 2., 1., 1.]),\n",
    "             ),\n",
    "            ),\n",
    "\n",
    "            # shift the reads all the way out of the window...\n",
    "            ('chr2L:68-76', 10,\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            yield check, kind, coord, shift_width, expected\n",
    "\n",
    "\n",
    "def test_local_coverage_read_strand():\n",
    "    \"\"\"\n",
    "    checks stranded full binning\n",
    "\n",
    "    excludes bigwig since strand doesn't make sense for that format.\n",
    "    \"\"\"\n",
    "    def check(kind, coord, read_strand, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord, read_strand=read_strand)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed']:\n",
    "        for coord, read_strand, expected in (\n",
    "            ('chr2L:1-20', '+',\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "            ('chr2L:1-20', '-',\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            yield check, kind, coord, read_strand, expected\n",
    "\n",
    "\n",
    "def test_local_coverage_fragment_size():\n",
    "    def check(kind, coord, fragment_size, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord, fragment_size=fragment_size)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed']:\n",
    "        for coord, fragment_size, expected in (\n",
    "            ('chr2L:1-20', 7,\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "\n",
    "            ('chr2L:68-76', 6,\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 1., 2., 2., 2., 2., 2., 1.]),\n",
    "             ),\n",
    "            ),\n",
    "\n",
    "            ('chr2L:68-76', 1,\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 0., 1., 0., 0., 0., 1., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            yield check, kind, coord, fragment_size, expected\n",
    "\n",
    "\n",
    "def test_local_coverage_score():\n",
    "    def check(kind, coord, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord, use_score=True)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bigbed', 'bed']:\n",
    "        for coord, expected in (\n",
    "            ('chr2L:1-20',\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10,   11,   12,   13,   14,   15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 255., 255., 255., 255., 255., 0., 0., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "            ('chr2L:68-76',\n",
    "             (\n",
    "                 np.array([68, 69, 70,   71,   72,   73,   74,   75]),\n",
    "                 np.array([0., 0., 510., 510., 510., 510., 510., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            yield check, kind, coord, expected\n",
    "\n",
    "\n",
    "def test_local_coverage_full():\n",
    "    \"\"\"generator of tests for local coverage\n",
    "\n",
    "    ensures that all formats are consistent in their results when retrieving\n",
    "    the full un-binned data.\n",
    "    \"\"\"\n",
    "    def check(kind, coord, processes, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord, processes=processes)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            ('chr2L:1-20',\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "            ('chr2L:68-76',\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 0., 2., 2., 2., 2., 2., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "            ('chr2L:568-576',\n",
    "             (\n",
    "                 np.array([568, 569, 570, 571, 572, 573, 574, 575]),\n",
    "                 np.array([0., 0.,   0.,  0.,  0.,  0.,  0.,  0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            for processes in [None, PROCESSES]:\n",
    "\n",
    "                yield check, kind, coord, processes, expected\n",
    "\n",
    "\n",
    "def test_local_coverage_binned():\n",
    "    \"\"\"generator of tests for local coverage\n",
    "\n",
    "    ensures that all formats are consistent in their results when retrieving\n",
    "    binned data.\n",
    "    \"\"\"\n",
    "    def check(kind, coord, processes, expected):\n",
    "        if kind == 'bigwig':\n",
    "            result = gs[kind].local_coverage(coord, bins=8, method='get_as_array', processes=processes)\n",
    "        else:\n",
    "            try:\n",
    "                result = gs[kind].local_coverage(coord, bins=8, processes=processes)\n",
    "            except NotImplementedError:\n",
    "                raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        try:\n",
    "            assert np.allclose(result[0], expected[0]) and np.allclose(result[1], expected[1])\n",
    "        except:\n",
    "            print (kind, coord, result, expected)\n",
    "            raise\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            ('chr2L:1-20',\n",
    "             (\n",
    "                 np.array([ 1., 3.57142857, 6.14285714, 8.71428571, 11.28571429, 13.85714286, 16.42857143, 19.]),\n",
    "                 np.array([ 0., 0.,         0.,         0.,         1.,          1.,          0.,          0. ]),\n",
    "             ),\n",
    "            ),\n",
    "            ('chr2L:68-76',\n",
    "             (\n",
    "                 np.array([68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 0., 2., 2., 2., 2., 2., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            for processes in [None, PROCESSES]:\n",
    "                yield check, kind, coord, processes, expected\n",
    "\n",
    "\n",
    "def test_array_binned():\n",
    "    def check(kind, coord, processes, expected):\n",
    "        if kind == 'bigwig':\n",
    "            result = gs[kind].array(coord, bins=8, method='get_as_array', processes=processes)\n",
    "        else:\n",
    "            try:\n",
    "                result = gs[kind].array(coord, bins=8, processes=processes)\n",
    "            except NotImplementedError:\n",
    "                raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        try:\n",
    "            assert np.allclose(result, expected)\n",
    "        except:\n",
    "            print (kind, coord, result, expected)\n",
    "            raise\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            (['chr2L:1-20'],\n",
    "             np.array([[0., 0., 0., 0., 1., 1., 0., 0. ]]),\n",
    "            ),\n",
    "            (['chr2L:1-20', 'chr2L:1-20[-]'],\n",
    "             np.array([[0., 0., 0., 0., 1., 1., 0., 0. ],\n",
    "                       [0., 0., 1., 1., 0., 0., 0., 0. ]]),\n",
    "            ),\n",
    "            (['chr2L:68-76'],\n",
    "             np.array([[0., 0., 2., 2., 2., 2., 2., 0.]]),\n",
    "             ),\n",
    "        ):\n",
    "            for processes in [None, PROCESSES]:\n",
    "                yield check, kind, coord, processes, expected\n",
    "\n",
    "\n",
    "def test_array_binned_preserve_total():\n",
    "    def check(kind, coord, processes, expected):\n",
    "        kwargs = dict(features=coord, bins=8, processes=processes, preserve_total=True)\n",
    "        if kind == 'bigwig':\n",
    "            assert_raises(ArgumentError, gs[kind].array, method='get_as_array', **kwargs)\n",
    "            return\n",
    "        else:\n",
    "            try:\n",
    "                result = gs[kind].array(**kwargs)\n",
    "            except NotImplementedError:\n",
    "                raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        try:\n",
    "            assert np.allclose(result, expected)\n",
    "        except:\n",
    "            print (kind, coord, result, expected)\n",
    "            raise\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            (['chr2L:1-20'],\n",
    "             np.array([[0., 0., 0., 0., .5, .5, 0., 0. ]]),\n",
    "            ),\n",
    "            (['chr2L:1-20', 'chr2L:1-20[-]'],\n",
    "             np.array([[0., 0., 0., 0., .5, .5, 0., 0. ],\n",
    "                       [0., 0., .5, .5, 0., 0., 0., 0. ]]),\n",
    "            ),\n",
    "            (['chr2L:68-76'],\n",
    "             np.array([[0., 0., .4, .4, .4, .4, .4, 0.]]),\n",
    "             ),\n",
    "        ):\n",
    "            for processes in [None, PROCESSES]:\n",
    "                yield check, kind, coord, processes, expected\n",
    "\n",
    "\n",
    "def test_invalid_arguments():\n",
    "    def check(kind, kw):\n",
    "        assert_raises(ArgumentError, gs[kind].array, **kw)\n",
    "\n",
    "    default_kwargs = dict(features='chr2L:1-20', bins=8)\n",
    "    # List the kinds of args that should return ArgumentErrors.\n",
    "    invalid_kwargs = {\n",
    "        'bam': (\n",
    "            dict(use_score=True),\n",
    "        ),\n",
    "\n",
    "        'bigwig': (\n",
    "            dict(read_strand=True),\n",
    "            dict(fragment_size=100),\n",
    "            dict(shift_width=100),\n",
    "            dict(use_score=True),\n",
    "            dict(preserve_total=True),\n",
    "        ),\n",
    "    }\n",
    "    for kind, kwarg_list in sorted(invalid_kwargs.items()):\n",
    "        for kwargs in kwarg_list:\n",
    "            default_kwargs_copy = default_kwargs.copy()\n",
    "            default_kwargs_copy.update(kwargs)\n",
    "            yield check, kind, default_kwargs_copy\n",
    "\n",
    "\n",
    "def test_array_ragged():\n",
    "    def check(kind, coord, processes, expected):\n",
    "        if kind == 'bigwig':\n",
    "            result = gs[kind].array(coord, method='get_as_array', ragged=True, processes=processes)\n",
    "        else:\n",
    "            try:\n",
    "                result = gs[kind].array(coord, processes=processes, ragged=True)\n",
    "            except NotImplementedError:\n",
    "                raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        try:\n",
    "            if isinstance(result, list):\n",
    "                for i, j in zip(result, expected):\n",
    "                    assert np.allclose(i, j)\n",
    "        except:\n",
    "            print (kind, coord, result, expected)\n",
    "            raise\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            (['chr2L:1-20'],\n",
    "             np.array([[0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,   1.,  0.,  0.,  0.,  0.,  0.]])\n",
    "            ),\n",
    "            (\n",
    "                [['chr2L:1-20'], ['chr2L:1-19']],\n",
    "                [\n",
    "                    np.array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,   1.,  0.,  0.,  0.,  0.,  0.]),\n",
    "                    np.array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,   1.,  0.,  0.,  0.,  0., ])\n",
    "                ]\n",
    "            ),\n",
    "\n",
    "            ([['chr2L:68-76', 'chr2L:68-76'], 'chr2L:68-76'],\n",
    "             [\n",
    "                 np.array([0., 0., 2., 2., 2., 2., 2., 0., 0., 0., 2., 2., 2., 2., 2., 0.]),\n",
    "                 np.array([0., 0., 2., 2., 2., 2., 2., 0.])\n",
    "             ],\n",
    "            ),\n",
    "        ):\n",
    "            for processes in [None, PROCESSES]:\n",
    "                yield check, kind, coord, processes, expected\n",
    "\n",
    "\n",
    "def test_bigwig_methods():\n",
    "    \"\"\"\n",
    "    using bx-python's `summarize` gives different results than UCSC's summarize\n",
    "    and summarizing from the original BAM.\n",
    "\n",
    "    possible bx-python bug, or just a resolution thing?  Haven't worked that\n",
    "    out yet, so test results here.\n",
    "    \"\"\"\n",
    "    # y0 is identical to the above test, and should be the same as\n",
    "    # local_coverage for bam, bigbed, and bed as well.\n",
    "    #\n",
    "    # y1 may not always be identical . . . I think it's due to resolution of\n",
    "    # the summarized file, especially when working with such short genomic\n",
    "    # intervals?\n",
    "    x0 = np.array([1., 3.57142857, 6.14285714, 8.71428571, 11.28571429, 13.85714286, 16.42857143, 19.])\n",
    "    y0 = np.array([ 0., 0.,         0.,         0.,         1.,          1.,          0.,          0. ]),\n",
    "    y1 = np.array([ 0., 0.,         0.,         0.,         1.,          1.,          1.,          0. ]),\n",
    "    #                                                         note difference here ---^\n",
    "    x, y = gs['bigwig'].local_coverage('chr2L:1-20', bins=8, method='get_as_array')\n",
    "    assert np.allclose(x0, x)\n",
    "    assert np.allclose(y0, y)\n",
    "\n",
    "    x, y = gs['bigwig'].local_coverage('chr2L:1-20', bins=8, method='summarize')\n",
    "    assert np.allclose(x0, x)\n",
    "    assert np.allclose(y1, y)\n",
    "\n",
    "    x, y = gs['bigwig'].local_coverage('chr2L:1-20', bins=8, method='ucsc_summarize')\n",
    "    assert np.allclose(x0, x)\n",
    "    assert np.allclose(y0, y)\n",
    "\n",
    "\n",
    "def test_nonbigwig_kwargs():\n",
    "    \"\"\"\n",
    "    test non-bigwig kwargs with a bigwig file\n",
    "\n",
    "    these kwargs are not valid for bigwigs because they don't make sense for\n",
    "    the format.\n",
    "    \"\"\"\n",
    "    nonwigs = {\n",
    "        'read_strand': '+',\n",
    "        'fragment_size': 200,\n",
    "        'shift_width': 80,\n",
    "        'use_score': True,\n",
    "    }\n",
    "    for nw in nonwigs.items():\n",
    "        kwargs = dict((nw,))\n",
    "        try:\n",
    "            assert_raises(\n",
    "                ArgumentError, gs['bigwig'].local_coverage, 'chr2L:1-20',\n",
    "                **kwargs)\n",
    "        except AssertionError:\n",
    "            print (kwargs)\n",
    "            raise\n",
    "\n",
    "    assert_raises(ArgumentError, gs['bigwig'].local_coverage, 'chr2L:1-20',\n",
    "                  method='ucsc_summarize', preserve_total=True)\n",
    "\n",
    "\n",
    "def test_local_coverage_kwarg_errors():\n",
    "    \"\"\"\n",
    "    ensures the argument-checking is working\n",
    "    \"\"\"\n",
    "    def check(kind, kwargs):\n",
    "        assert_raises(\n",
    "            ArgumentError,\n",
    "            gs[kind].local_coverage, **kwargs)\n",
    "\n",
    "    items = [\n",
    "        ('bam', dict(bins='a', features='chr2L:1-20')),\n",
    "        ('bam', dict(use_score=True, features='chr2L:1-20')),\n",
    "        ('bam', dict(bins=[8, 10], features=['chr2L:1-20'])),\n",
    "\n",
    "    ]\n",
    "\n",
    "    for kind, kwargs in items:\n",
    "        yield check, kind, kwargs\n",
    "\n",
    "\n",
    "def test_local_coverage_full_multifeature():\n",
    "    def check(kind, coord, expected):\n",
    "        try:\n",
    "            result = gs[kind].local_coverage(coord)\n",
    "        except NotImplementedError:\n",
    "            raise SkipTest(\"Incompatible bx-python version for bigBed\")\n",
    "        assert np.all(result[0] == expected[0]) and np.all(result[1] == expected[1]), (kind, coord, result)\n",
    "\n",
    "    for kind in ['bam', 'bigbed', 'bed', 'bigwig']:\n",
    "        for coord, expected in (\n",
    "            (['chr2L:1-20', 'chr2L:68-76'],\n",
    "             (\n",
    "                 np.array([1,  2,  3,  4,  5,  6,  7,  8,  9,  10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 68, 69, 70, 71, 72, 73, 74, 75]),\n",
    "                 np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 2., 2., 2., 2., 2., 0.]),\n",
    "             ),\n",
    "            ),\n",
    "        ):\n",
    "            yield check, kind, coord, expected\n",
    "\n",
    "\n",
    "def test_errors():\n",
    "    \"these things should raise errors\"\n",
    "    def check(error, callable_obj, args, kwargs):\n",
    "        assert_raises(error, callable_obj, *args, **kwargs)\n",
    "\n",
    "    class X(metaseq.filetype_adapters.BaseAdapter):\n",
    "        def make_fileobj(self):\n",
    "            return None\n",
    "\n",
    "    items = [\n",
    "        (ValueError, metaseq.filetype_adapters.BaseAdapter, (metaseq.example_filename('gdc.bed'),), {}),\n",
    "        (NotImplementedError, metaseq.filetype_adapters.BigWigAdapter(metaseq.example_filename('gdc.bigwig')).__getitem__, (0,), {}),\n",
    "        (ValueError, X(\"\").__getitem__, (0,), {}),\n",
    "        #(ValueError, gs['bam'].local_coverage, ['chr2L:1-5', 'chr2L:1-5'], dict(processes=PROCESSES)),\n",
    "    ]\n",
    "    for error, callable_obj, args, kwargs in items:\n",
    "        yield check, error, callable_obj, args, kwargs\n",
    "\n",
    "def test_supported_formats():\n",
    "    assert set(metaseq._genomic_signal.supported_formats()) \\\n",
    "        == set(['bam', 'bigwig', 'bed', 'gff', 'gtf', 'vcf', 'bigbed'])\n",
    "    assert_raises(ValueError, metaseq.genomic_signal, \"\", 'unsupported')\n",
    "\n",
    "def test_bam_genome():\n",
    "    assert gs['bam'].genome() == {'chr2L', (0, 23011544L)}\n",
    "\n",
    "def test_bam_mmr():\n",
    "    import os\n",
    "    assert gs['bam'].mapped_read_count(force=True) == 8\n",
    "    gs['bam']._readcount = None\n",
    "    assert gs['bam'].mapped_read_count(force=False) == 8, gs['bam'].mapped_read_count(force=False)\n",
    "\n",
    "def test_bigwig_out_of_range():\n",
    "    x, y = gs['bigwig'].local_coverage('chr1:1-100', bins=None)\n",
    "    assert y.sum() == 0\n",
    "    x, y = gs['bigwig'].local_coverage('chr1:1-100', bins=None, method='ucsc_summarize')\n",
    "    assert y.sum() == 0\n",
    "\n",
    "    x, y = gs['bigwig'].local_coverage('chr2L:1-100', bins=None, method='ucsc_summarize')\n",
    "\n",
    "def test_valid_bigwig_method():\n",
    "    assert_raises(ValueError, gs['bigwig'].local_coverage, 'chr2L:1-100', method='nonexistent')\n",
    "\n",
    "def test_bigwig_zero_inf_nan():\n",
    "    def check(coord, kwargs, expected):\n",
    "        expected = [\n",
    "                np.array([ 1., 23., 45., 67., 89., 111., 133., 155.,  177., 199.]),\n",
    "                expected\n",
    "        ]\n",
    "        result = gs['bigwig'].local_coverage(coord, bins=10, **kwargs)\n",
    "\n",
    "        valid = ~(np.isinf(expected[1]) | np.isnan(expected[1]))\n",
    "        print(coord, kwargs, result)\n",
    "        assert np.all(result[0] == expected[0])\n",
    "        assert np.all(np.isinf(expected[1]) == np.isinf(result[1]))\n",
    "        assert np.all(np.isnan(expected[1]) == np.isnan(result[1]))\n",
    "        assert np.all(result[1][valid] == expected[1][valid])\n",
    "\n",
    "\n",
    "    for coord, kwargs, expected in [\n",
    "\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=True, method='get_as_array'),\n",
    "            np.array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.])\n",
    "        ),\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=False, method='get_as_array'),\n",
    "            np.array([ nan,  nan,  nan,  nan,  nan,  nan,  nan,  nan,  nan,  nan])\n",
    "        ),\n",
    "\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=True, method='summarize', function='mean'),\n",
    "            np.array([ 1., 0. , 0.,  2.,  0.,  0.,   0.,   1.625, 1.,   0.   ])\n",
    "        ),\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=False, method='summarize', function='mean'),\n",
    "            np.array([1., nan, nan, 2.,  nan, nan,  nan,  1.625,  1.,    nan])\n",
    "        ),\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=False, method='summarize', function='std'),\n",
    "            np.array([ 1.,   nan, nan, 4.,  nan, nan,  nan,  2.875, 1.,    nan])\n",
    "        ),\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=True, method='summarize', function='std'),\n",
    "            np.array([ 1.,   0. , 0. , 4.,  0. , 0. ,  0. ,  2.875, 1.,    0. ])\n",
    "        ),\n",
    "\n",
    "        # ucsc_summarize has slightly different value for [-3] and always\n",
    "        # returns zero for missing values (even when zero_inf and zero_nan are\n",
    "        # False)\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=False, method='ucsc_summarize', function='mean'),\n",
    "            np.array([1., 0. , 0. , 2.,  0. , 0. ,  0.,   1.5  ,  1.,    0.])\n",
    "        ),\n",
    "\n",
    "        # ucsc_summarize disagrees with bx-python for std\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=False, method='ucsc_summarize', function='std'),\n",
    "            np.array([ 0.,   0. , 0. , 0.,  0. , 0. ,  0. ,  0.527046, 0.,    0. ])\n",
    "        ),\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=False, method='summarize', function='min'),\n",
    "            np.array([  1.,  inf,  inf,   2.,  inf,  inf,  inf,   1.,   1.,  inf])\n",
    "        ),\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=True, method='summarize', function='min'),\n",
    "            np.array([  1.,  0.,  0.,   2.,  0.,  0.,  0.,   1.,   1.,  0.])\n",
    "        ),\n",
    "\n",
    "        # ucsc_summarize *does* aggree on min...\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=False, method='ucsc_summarize', function='min'),\n",
    "            np.array([  1.,  0.,  0.,   2.,  0.,  0.,  0.,   1.,   1.,  0.])\n",
    "        ),\n",
    "\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=False, zero_nan=False, method='summarize', function='max'),\n",
    "            np.array([  1., -inf, -inf,   2., -inf, -inf, -inf,   2.,   1., -inf])\n",
    "        ),\n",
    "\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=True, method='summarize', function='max'),\n",
    "            np.array([  1., 0., 0.,   2., 0., 0., 0.,   2.,   1., 0.])\n",
    "        ),\n",
    "\n",
    "        # ...and on max\n",
    "        (\n",
    "            'chr2L:1-200',\n",
    "            dict(zero_inf=True, zero_nan=True, method='ucsc_summarize', function='max'),\n",
    "            np.array([  1., 0., 0.,   2., 0., 0., 0.,   2.,   1., 0.])\n",
    "        ),\n",
    "    ]:\n",
    "        yield check, coord, kwargs, expected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
